straxen/plugins/peak_processing.py

```diff
@@ -332,97 +334,283 @@ def find_n_competing(peaks, windows, fraction):
-        and check if it casts the largest shadow
+        For each peak in peaks, check if there is a shadow-casting peak
+        and check if it casts the first or second largest shadow
         """
-        for p_i, p_a in enumerate(peaks):
+        for p_i, (p_a, sigma) in enumerate(zip(peaks, sigmas)):
             # reset for every peak
-            new_shadow = 0
-            s2_indices = touching_windows[p_i]
-            for s2_idx in range(s2_indices[0], s2_indices[1]):
-                s2_a = pre_s2_peaks[s2_idx]
-                if p_a['center_time'] - s2_a['center_time'] <= 0:
+            indices = touching_windows[p_i]
+            shadow_ref_bk = 0
+            for idx in range(indices[0], indices[1]):
+                s_a = pre_peaks[idx]
+                near_dt = p_a['center_time'] - s_a['center_time']
+                if near_dt <= 0:
                     continue
-                new_shadow = s2_a['area'] * (
-                        p_a['center_time'] - s2_a['center_time'])**exponent
-                if new_shadow > res['shadow'][p_i]:
+                new_shadow = s_a['area'] * near_dt**exponent
+                if pos_corr:
+                    distance = np.sqrt((p_a['x'] - s_a['x']) ** 2 + (p_a['y'] - s_a['y']) ** 2)
+                    distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
+                    shadow_ref = new_shadow * 2 / (np.pi * sigma * (1 + (distance / sigma) ** 2))
+                else:
+                    shadow_ref = new_shadow
+                res['shadow_sum'][p_i] += shadow_ref
+                if near_dt < res['near_dt'][p_i]:
+                    res['near_dt'][p_i] = near_dt
+                if shadow_ref > shadow_ref_bk:
+                    shadow_ref_bk = shadow_ref
+                    res['alt_pre_x'][p_i] = res['pre_x'][p_i]
+                    res['alt_pre_y'][p_i] = res['pre_y'][p_i]
+                    res['pre_x'][p_i] = s_a['x']
+                    res['pre_y'][p_i] = s_a['y']
+                    res['alt_shadow'][p_i] = res['shadow'][p_i]
                     res['shadow'][p_i] = new_shadow
-                    res['pre_s2_area'][p_i] = s2_a['area']
-                    res['shadow_dt'][p_i] = p_a['center_time'] - s2_a['center_time']
-                    res['pre_s2_x'][p_i] = s2_a['x']
-                    res['pre_s2_y'][p_i] = s2_a['y']
+                    res['alt_pre_area'][p_i] = res['pre_area'][p_i]
+                    res['pre_area'][p_i] = s_a['area']
+                    res['alt_shadow_dt'][p_i] = res['shadow_dt'][p_i]
+                    res['shadow_dt'][p_i] = p_a['center_time'] - s_a['center_time']
+
+
+@export
+@strax.takes_config(
+    strax.Option('ambience_divide_time', default=False,
+                 help='Whether to divide area by time'),
+    strax.Option('ambience_divide_space', default=False,
+                 help='Whether to divide area by space'),
+    strax.Option(name='time_window_ambience', default=int(2e6),
```

Comment:
how this 2-msec value is optimized? Do you have any studies with different value?

Reply:
It is optimized for signal acceptance ~99% in ambience cut. I have not checked other values.